### PR TITLE
Two layer design with CAPSULE

### LINKED ISSUE #41: 2-layer design
During the 2021-04 MASQUE Interim, we discussed a design proposal from @bemasc to replace the one-layer flow ID design with a two-layer design. This design replaces the flow ID with two numbers: a stream ID and a context ID. The contents of the QUIC DATAGRAM frame would now look like:
```
HTTP/3 DATAGRAM Frame {
  Stream ID (i),
  Context ID (i),
  HTTP/3 Datagram Payload (..),
}
```

While the flow ID in the draft was per-hop, the proposal has better separation:
* the stream ID is per-hop, it maps to an HTTP request
* the context ID is end-to-end, it maps to context information inside that request

Intermediaries now only look at stream IDs and can be blissfully ignorant of context IDs.

In the room at the 2021-04 MASQUE Interim, multiple participants spoke in support of this design and no one raised objections. This issue exists to ensure that we have consensus on this change - please speak up if you disagree.

### LINKED ISSUE #42: Directionality of flow/context IDs
In draft-ietf-masque-h3-datagram-00, flow IDs are bidirectional. During the 2021-04 MASQUE Interim, we discussed the possibility of making them unidirectional. Here are the differences:

* Bidirectional
  * single shared namespace for both endpoints (even for client, odd for server)
  * it's easier to refer to peer's IDs, which makes negotiation easier

* Unidirectional
  * one namespace per direction
  * every bidirectional use needs to associate both directions somehow

I'm personally leaning towards keeping them bidirectional. Unlike QUIC streams, these do not have a state machine or flow control credits, so there is no cost to only using a single direction of a bidirectional flow for unidirectional use-cases.

### LINKED ISSUE #43: Ability to accept or reject registration of context IDs
During the 2021-04 MASQUE Interim, we discussed having the ability to accept or reject registration of context IDs. An important motivation for this is the fact that context ID registrations take up memory (for example it could be a compression context which contains the data that is elided) and therefore we need to prevent endpoints from arbitrarily making their peer allocate memory. While we could flow control context ID registration, a much simpler solution would be to have a way for endpoints to close/shutdown a registration.

### LINKED ISSUE #44: Pick mechanism to register context IDs (header vs message)
During the 2021-04 MASQUE Interim, we discussed multiple options for registering context IDs. Without going into the specific of the encoding, there are two broad classes of solutions:

* "Header" design
  * Once-at-setup registration
  * This happens during the HTTP request/response
  * Simplest solution is to use an HTTP header such as `Datagram-Flow-Id` (encoding TBD)
  * Example use-case: CONNECT-UDP without extensions
* "Message" design
  * Mid-stream registration
  * This can happen at any point during the lifetime of the request stream
  * This would involve some sort of "register" message (encoding TBD, see separate issue)
  * Example use-case: CONNECT-IP on-the-fly compression (this can't be done at tunnel setup time because the bits top be compressed aren't always known at that time)
 
It's possible to implement once-at-setup registration using the Message design, but it isn't possible to implement mid-stream registration. Therefore I think we should go with the Message design. It would also be possible to implement both but I don't think that provides much value.

### PR DESCRIPTION
**Pull request by DavidSchinazi**
This PR aims to take the design discussions we had at the last
interim, and propose a potential solution. This PR should in theory
contain enough information to allow folks to implement this new
design. That will allow us to perform interop testing and confirm
whether the design is sound.

[An HTML version of the draft with these changes is available here.](https://ietf-wg-masque.github.io/draft-ietf-masque-h3-datagram/two_layer_capsule/draft-ietf-masque-h3-datagram.html)

Closes #41, closes #42, closes #43, closes #44.

### CODE REVIEW

File: draft-ietf-masque-h3-datagram.md
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:154]: This requirement seems awkard to implement as specified. I think some of that comes from mixing length checks and parsing. To elabortate:


* it states endpoints, but what about intermediaries (they need to parse just the Quarter Stream ID but should they forward a frame whose length is exactly the Quarter Stream ID?)
* for an endpoint parsing both fields it probably needs to be an 'and' 
* since the fields are varints its harder to check the DATAGRAM payload length can satisfy the field requirements. The obvious case is if payload length is 1 (two minimally encoded varints) but other combos make it harder.

The way that H3 specs the requirement isn't a perfect fit but I think the spirit is more closely align with what you're trying to say - https://tools.ietf.org/html/draft-ietf-quic-http-34#section-7.1

Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Good point, one of those is hop-by-hop but the other is end-to-end. I've tweaked the requirements here to reflect that.

File: draft-ietf-masque-h3-datagram.md
Comment (tfpauly) [inline draft-ietf-masque-h3-datagram.md:170]: Bikeshed bikeshed bikeshed...

File: draft-ietf-masque-h3-datagram.md
Comment (tfpauly) [inline draft-ietf-masque-h3-datagram.md:272]:
```suggestion
unilaterally informs its peer of the closure. Endpoints can use
```

File: draft-ietf-masque-h3-datagram.md
Comment (tfpauly) [inline draft-ietf-masque-h3-datagram.md:274]:
```suggestion
registered by either themselves, or by their peer. Endpoints MAY use the
```

**Review by tfpauly — APPROVED** [review]

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:58]: Indeed :)

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Ah, that was my original intention. Thanks. Fixed.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Done.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Reworded to use divisible instead of modulo

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:170]: Nope nope nope

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Done

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Done

**Review by DavidSchinazi — COMMENTED** [review]
Thanks for the review!

File: draft-ietf-masque-h3-datagram.md
Comment (bemasc) [inline draft-ietf-masque-h3-datagram.md:274]: Allowing endpoints to close contexts created by the peer seems confusing to me.  If my peer just closed the context I've been using ... what do I do?  Do I reopen it?  Do I fail the connection?

File: draft-ietf-masque-h3-datagram.md
Comment (bemasc) [inline draft-ietf-masque-h3-datagram.md:275]: I think it would probably be better to distinguish CLOSE from "registration rejected".  Did the server close the context because it accepted my packets and believes the context is no longer needed?  Or did it close because I hit a resource limit?  Or was my request incompatible?

If something can be used to communicate failures, then I think it probably needs an error code, so the recipient knows what to do next.

File: draft-ietf-masque-h3-datagram.md
Comment (bemasc) [inline draft-ietf-masque-h3-datagram.md:172]:
```suggestion
CAPSULE (placeholder, final name TBD) allows reliably sending request-related information end-to-end, even in
```

**Review by bemasc — APPROVED** [review]
I think this is a good change.  Thanks!

There are some things that I think we should still discuss, but we shouldn't hold up this change on those details.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:274]: I'd say that depends on the application. In CONNECT-UDP if you close the main context then things break, so you might as well close the stream. But as another example, if your peer introduced a CONNECT-IP compression context and used it for a while and you're seeing low usage of it and want to reclaim memory, you can close than context and have the tunnel still be functional. Let's move this conversation to #43.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:275]: That's an interesting idea. I've filed #53 to track it so we don't forget after this is merged.

File: draft-ietf-masque-h3-datagram.md
Comment (afrind) [inline draft-ietf-masque-h3-datagram.md:535]: TImestamp -> Timestamp

File: draft-ietf-masque-h3-datagram.md
Comment (afrind) [inline draft-ietf-masque-h3-datagram.md:434]: I assume this is for the 'extension-string' parameter in Register Datagram Context.  The word key first appears here.

File: draft-ietf-masque-h3-datagram.md
Comment (afrind) [inline draft-ietf-masque-h3-datagram.md:170]: After reading this document, the main benefit I see to CAPSULE is that it makes it easier to deploy future end-to-end extension "frames" without having to modify intermediaries (because H3 drops unknown frames).  It may also make it easier to extend in client <--> server cases where the H3 stack itself is not easily modifiable (owned by a third-party, say).

Eventually it would probably warrant its own document and not be buried in DATAGRAM?

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Added more text to explain contexts

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Done

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:113]: The next sentence explains this further, I prefer initiated over defined.

File: draft-ietf-masque-h3-datagram.md
Comment (afrind) [inline draft-ietf-masque-h3-datagram.md:157]: A generic HTTP library capable of receiving and sending datagrams could be used to build either an intermediary or an endpoint.  WIth this specification, It would need two sets of APIs, one that had the context ID (for endpoints) and one without (for intermediaries).  If datagrams carry context IDs then intermediaries should be able to parse them.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:117]: I find used to be clearer

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:157]: Done

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:181]: I agree with your benefit analysis. I'd prefer to avoid splitting the document further though. Let's wait and see if anyone wants to use CAPSULE but cannot implement DATAGRAM.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Done

Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Capsules are by definition end-to-end. If intermediaries start looking inside, they might cause ossification. In theory, I'd happily use end-to-end encryption to protect them but we don't have keys for this purpose.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Done

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:227]: I consider that to be a feature rather than a bug. If we don't exercise this path, implementors will cut corners and we won't be able to deploy extensions down the road. Requiring the main path to use the extension mechanism ensures that the mechanism is exercised and functional.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:253]: Done

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Done

File: draft-ietf-masque-h3-datagram.md
Comment (LPardue) [reply draft-ietf-masque-h3-datagram.md:177]: I think adding a secure integrity mechanism is overkill. I'd treat these more like how headers and intermediaries work. Most intermediaries really have no business looking at the stuff, but some will decide to do stuff and if they break things then hopefully the failure to abide the spec will be reflected in market position.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: These are the keys from that section, I added references.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: Done

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:162]: Your HTTP implementation is going to need to know whether to act on datagrams (endpoint) or forward them (intermediaries). Plumbing that bool to the parse function doesn't seem complex?

Comment (LPardue) [reply draft-ietf-masque-h3-datagram.md:157]: What would intermediaries do with it? It might be fair to parse it and tell your API users "this field is entirely meaningless unless you are an endpoint".

**Review by DavidSchinazi — COMMENTED** [review]
Thanks for the review!

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:113]: Odd-numbered context IDs allow servers to register their own contexts. For example, a CONNECT-IP server could choose to start compressing using a compression context of its choosing.

File: draft-ietf-masque-h3-datagram.md
Comment (afrind) [reply draft-ietf-masque-h3-datagram.md:162]: > What would intermediaries do with it?

The same thing it does with other opaque things it get from the peer (eg: body bytes, datagram payloads) - forward to the other peer.

Whatever thing handles an HTTP stream would now have APIs like

```
onDatagram(payload)
sendDatagram(payload)
```
for intermediaries, or, if you are for endpoints:

```
onDatagram(contextID, payload)
sendDatagram(contextID, payload)
```

I'd prefer to have the library only have the second variant.  If using the library to implement a proxy, then the proxy implementation would pass the context ID and payload from one peer's `onDatagram` to the other side's `sendDatagram`. 

Logging/stats are another possible use.

I guess the broader point is -- as this spec is written, the datagram has to have that context ID, so why be so prescriptive about what the intermediary can and can't do with it.

File: draft-ietf-masque-h3-datagram.md
Comment (afrind) [reply draft-ietf-masque-h3-datagram.md:227]: Most wg's approach grease by defining it and encouraging implementers to use it liberally enough to prevent this, but not require it.  HTTP/3 doesn't require a GREASE frame in order to function.

File: draft-ietf-masque-h3-datagram.md
Comment (LPardue) [reply draft-ietf-masque-h3-datagram.md:113]: Yeah this makes sense, I just didn't digest all of the concepts before posting.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:162]: We're considering making the Context ID optional (see #45). If intermediaries start parsing the context ID, it'll prevent us from doing that.

File: draft-ietf-masque-h3-datagram.md
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:227]: This isn't about grease, it's about using your extension mechanism. HTTP/3 forces all HTTP implementations to correctly support HTTP headers, because otherwise they couldn't send or parse `:method`. That guarantees that everyone correctly handles headers. We could have said that if no HEADERS frame is sent then it means GET, but we decided against that.

File: draft-ietf-masque-h3-datagram.md
Comment (afrind) [reply draft-ietf-masque-h3-datagram.md]: I was being a bit tongue in cheek by suggesting security measures here. It strikes me as overkill too.

So is it explicitly forbidden for a proxy to register its own context ID on a stream via CAPSULE?  This would require some careful context-ID allocations by the proxy to avoid potential collisions, which in turn break the rules about the context-IDs being end-to-end.

I don't have a use case, but I'm not sure I see the the value in locking this down quite so tightly.  Are there designs that depend on the context identifier being the same at every hop?  Perhaps tunneling QUIC in 'forward' mode relies on this?

Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: The core tenet of the two-layer design is that Stream IDs are per-hop and context IDs are end-to-end. That allows extensions to rely on the fact that both endpoints have the same view of context IDs. CAPSULE exists as an end-to-end mechanism, and similarly relies on that property. If we need a new per-hop mechanism, then we can use a different h3 frame. I've clarified that intermediaries don't send capsules apart from forwarding them, so they're not allowed to inject their own context IDs.

File: draft-ietf-masque-h3-datagram.md
Comment (afrind) [reply draft-ietf-masque-h3-datagram.md:162]: Whatever mechanism the endpoint will use to know if it should parse the context ID can also be used by the intermediary as well.  My point was that in this PR, they are required.

Presumably if/when negotiation is added the specification can be updated to reflect that.

Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:162]: I've filed #54 to keep track of this discussion so it doesn't get lost when this is merged. Let's continue this conversation there.

**Comment by DavidSchinazi** [comment]
@vasilvv you're referring to #45 which this PR isn't trying to address. That'll happen in a subsequent PR.

**Review by LPardue — APPROVED** [review]

**Pull request merged**
